[lldb] Add priority support to synthetic frame providers

medismailben opened the pull request:
This patch adds `get_priority()` support to synthetic frame providers to enable priority-based selection when multiple providers match a thread. This is the first step toward supporting frame provider chaining for visualizing coroutines, Swift async tasks, and et al.

Priority ordering follows Unix nice convention where lower numbers indicate higher priority (0 = highest). Providers without explicit priority return `std::nullopt`, which maps to UINT32_MAX (lowest priority), ensuring backward compatibility with existing providers.

The implementation adds `GetPriority()` as a virtual method to `SyntheticFrameProvider` base class, implements it through the scripting interface hierarchy (`ScriptedFrameProviderInterface` and `ScriptedFrameProviderPythonInterface`), and updates `Thread::GetStackFrameList()` to sort applicable providers by priority before attempting to load them.

Python frame providers can now specify priority:

```python
@staticmethod
def get_priority():
   return 10  # Or return None for default priority.
```

bulbazord commented on lldb/source/Target/Thread.cpp:
```diff
@@ -1455,16 +1455,34 @@ StackFrameListSP Thread::GetStackFrameList() {
-                           "Failed to load scripted frame provider: {0}");
-          }
-          break; // Use first matching descriptor (success or failure).
+          applicable_descriptors.push_back(&descriptor);
```
nit: one-line if blocks shouldn't have braces.

bulbazord reviewed (commented):
Overall idea looks good to me. Left a few inline comments.

medismailben commented on lldb/source/Target/Thread.cpp:
```diff
@@ -1455,16 +1455,34 @@ StackFrameListSP Thread::GetStackFrameList() {
+
+      // Load the highest priority provider that successfully instantiates.
+      for (const auto *descriptor : applicable_descriptors) {
+        if (llvm::Error error = LoadScriptedFrameProvider(*descriptor)) {
```
Not really, descriptors just contains, the name of the class to load, the argument dictionary to pass to it and the priority. Loading them creates the python object, stores it etc ... Changing the other would be actually be more expensive.

medismailben committed:
[lldb] Add priority support to synthetic frame providers

This patch adds `get_priority()` support to synthetic frame providers
to enable priority-based selection when multiple providers match a thread.
This is the first step toward supporting frame provider chaining for
visualizing coroutines, Swift async tasks, and et al.

Priority ordering follows Unix nice convention where lower numbers
indicate higher priority (0 = highest). Providers without explicit
priority return `std::nullopt`, which maps to UINT32_MAX (lowest priority),
ensuring backward compatibility with existing providers.

The implementation adds `GetPriority()` as a virtual method to
`SyntheticFrameProvider` base class, implements it through the scripting
interface hierarchy (`ScriptedFrameProviderInterface` and
`ScriptedFrameProviderPythonInterface`), and updates `Thread::GetStackFrameList()`
to sort applicable providers by priority before attempting to load them.

Python frame providers can now specify priority:

```python
@staticmethod
def get_priority():
   return 10  # Or return None for default priority
```

Signed-off-by: Med Ismail Bennani <ismail@bennani.ma>
